Add Geocoder

sunnymui opened the pull request:
Adds a geocoder module to the create post form to grab the longitude and latitude automatically from the LocationIQ api based on the street address entered in the form.

* Geocoder module logic is contained in `utilities/api.js`
* Button to activate geocoding added to `compositions/PostMarkup.js` and markup integrated in editAddressMarkup. Some props added to AddressMarkup for the geocoding button
* `Adminform.js` contains geocode interaction handler and an addition to the state, ApiLoading which tracks if the api is still loading the results. It disables the geocode button while results are still loading then reenables it when results have been loaded.
* Map display added for when longitude/latitude is entered. **Needs debugging** as the map only sets the correct view for the initial geocode activation. You have to erase the longitude or latitude fields to get the map to update again. Manual longitude/latitude entry also does not trigger map display.

Planning to also show a loading animation while ApiLoading = true in another update.

UPDATE
Issues resolved, see latest post

JMStudiosJoe commented on src/Utilities/api.js:
```diff
@@ -0,0 +1,95 @@
+class Geocoder {
```
this can likely go in as a service.

sunnymui replied:
In /src/api/api.js?

sunnymui replied:
changed locations to services/locationiq/api.js

JMStudiosJoe commented on src/Utilities/api.js:
```diff
@@ -0,0 +1,95 @@
+    };
+  }
+
+  _checkResponseJSON(response) {
```
is the _ needed? javascript doesnt really follow that practice anymore and can be confusing to other developers without comments but if it is not needed then best to not use it.

sunnymui replied:
Oh so I don't have to mark those kinds of internally used methods for a class?

sunnymui replied:
Refactored it like 7 times to make it more of a functional style. I went down a real rabbit hole rewriting this and with all the further reading I had to do.

sunnymui commented:
Main question is how I can get that map display to update when the longitude/latitude is changed. I'm not exactly sure how to use `useEffect` with the mapDisplay component

sunnymui commented:
I did some super serial refactoring and issue resolving based on the notes taken during the last Code for San Jose Meeting. These changes were implemented in the latest 7 commits added to this PR.

- Locationiq api module moved from utilities/api.js to the newly create services directory--services/locationiq/api.js
- API key removed from hardcoded implementation to a react local environmental variable accessed with the key `process.env.REACT_APP_LOCATION_IQ_KEY`. this api key is contained in .env.local, which has been added to the .gitignore file so it isn't shared publicly. Instructions on getting an api key for lcoationiq and creating your own .env.local file to get the app running added to the .env file. Should update the readme with this info.
- Locationiq module refactored several times, making it more functional and more in line with modern JS--async await functions, more destructuring, template literals, modularity. Way more modular now so if you don't want to use the preset getter functions, you can customize your own fairly easily. Also easy to add other getter functions for other endpoints. Also was able to remove some extraneous error checking. docs are more extensive too.
- Refactor/simplify the geocoder handler and logic in adminform.js to use the updated module and use async/await patterns
- Refactor the OSM mapdisplay.js component to be a little more organized and es6. Added a componentdidupdate lifecycle method so the preview map shown in the admin form gets updated on the fly as the state changes. Makes it a lot easier to tweak the positioning of a disaster visually. Also added a marker display if the radius is 0--makes more sense if a disaster is a single point. Added autofitting of the map bounds to the circle radius too.

JMStudiosJoe commented on src/components/MapDisplay/MapDisplay.js:
```diff
@@ -3,35 +3,103 @@ import React, { Component } from 'react'
+
     componentDidMount () {
-        const {
+        let {
```
these should stay as const no need for let.

sunnymui replied:
Oops I dunno why I changed that. I'll change it back

JMStudiosJoe commented on src/components/MapDisplay/MapDisplay.js:
```diff
@@ -3,35 +3,103 @@ import React, { Component } from 'react'
+  }
+};
 
 class MapDisplay extends Component {
```
this file might be best as a hook and useEffect

JMStudiosJoe replied:
but that can be for another branch

sunnymui replied:
It was already written with the older componentDidMount style so I just added the update lifecycle method in the same way. I still need to read up more on using hooks and useEffect so I decided I'll just leave that for a future update.

JMStudiosJoe commented on src/services/locationiq/api.js:
```diff
@@ -0,0 +1,97 @@
+  return `${endpoint}&lat=${lat}&lon=${lng}&format=json`
+}
+
+const responseToJSON = (response) => {
```
this might be overkill, we should be able to use the api.js instead of needing to use a combination of await and promises

sunnymui replied:
Do you mean using the makeRequest function in src/api/api.js? 

I guess I just wrote it all this way to try and make it a kinda portable encapsulated module.
